misc/vex/vex.go

```diff
@@ -0,0 +1,434 @@
+	}
+
+	// Save the VEX document for the current tag
+	f, err := os.Create(fmt.Sprintf("vex-%s.json", tag))
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create VEX file: %w", err)
+	}
+	defer f.Close()
+
+	e := json.NewEncoder(f)
+	e.SetIndent("", "  ")
+	if err = e.Encode(vexDoc); err != nil {
+		return nil, nil, fmt.Errorf("failed to write VEX file: %w", err)
+	}
+
+	// Run govulncheck and generate JSON result
+	// Need to generate JSON as well as OpenVEX for the following reasons:
+	//   - Subcomponent
+	//      - OpenVEX from govulncheck doesn't fill in subcomponents.
+	//   - Status
+	//      - govulncheck uses "not_affected" for all vulnerabilities. Need to determine "fixed" vulnerabilities.
+	//        cf. https://github.com/golang/go/issues/68338
+	findings, err := generateJSON(ctx)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to run govulncheck: %w", err)
+	}
+
+	product := *productID // Clone Trivy PURL
+	product.Version = tag
+	notAffected := make(map[UniqueKey]vex.Statement)
+	affected := make(map[UniqueKey]struct{})
+
+	// Update VEX document generated by govulncheck
+	for _, stmt := range vexDoc.Statements {
+		finding, ok := findings[stmt.Vulnerability.Name]
+		if !ok {
+			// Considered as "fixed" vulnerabilities
+			// cf. https://github.com/golang/go/issues/68338
+			continue
+		} else if len(finding.Finding.Trace) == 0 {
+			continue
+		}
+
+		namespace, name := path.Split(finding.Finding.Trace[0].Module)
+		subcomponent := &packageurl.PackageURL{
+			Type:      packageurl.TypeGolang,
+			Namespace: namespace,
+			Name:      name,
+		}
+
+		key := UniqueKey{
+			VulnerabilityID: stmt.Vulnerability.Name,
+			SubcomponentID:  subcomponent.String(),
+		}
+
+		if stmt.Status == vex.StatusAffected {
+			affected[key] = struct{}{}
+			continue
+		} else if stmt.Status != vex.StatusNotAffected {
+			continue
+		}
+
+		// Update the statement with product and subcomponent information
+		stmt.Products = []vex.Product{
+			{
+				// Fill in components manually
+				// cf. https://github.com/golang/go/issues/68152
+				Component: vex.Component{
+					ID: product.String(),
+					Identifiers: map[vex.IdentifierType]string{
+						vex.PURL: product.String(),
+					},
+				},
+				Subcomponents: []vex.Subcomponent{
+					{
+						Component: vex.Component{
+							ID: key.SubcomponentID,
+							Identifiers: map[vex.IdentifierType]string{
+								vex.PURL: key.SubcomponentID,
+							},
+						},
+					},
+				},
+			},
+		}
+		notAffected[key] = stmt
+	}
+
+	return notAffected, affected, nil
+}
+
+// generateVEX runs govulncheck with OpenVEX format and parses the output
+func generateVEX(ctx context.Context) (*vex.VEX, error) {
+	buf, err := runGovulncheck(ctx, "openvex")
+	if err != nil {
+		return nil, fmt.Errorf("failed to run govulncheck: %w", err)
+	}
+
+	vexDoc, err := vex.Parse(buf.Bytes())
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse govulncheck output: %w", err)
+	}
+	return vexDoc, nil
+}
```

Comment:
This function uses 7.5GB memory:
![Screenshot from 2024-07-09 15-18-59](https://github.com/aquasecurity/trivy/assets/91113035/e340cf42-a53f-448c-aafd-77e1e446c184)
looks like `vuln` package contain memory leak.

Reply:
I think the point is that a large repository is being scanned. https://github.com/golang/go/issues/56391

Reply:
They build a call graph. It makes sense to consume a lot of memory.

Reply:
Got it, thanks for the clarification.